### ISSUE TITLE
TpetraCrsGraph test coverage 

### ISSUE THREAD

**Issue by aprokop** [trilinos/ForTrilinos#135]

**Comment by kruger** [comment]
This is being addressed by PR #262.

Notes on minor issues with latest push (as of 04-24-2019).

Tests that do not work:
```
  !TODO-FIX ADD_SUBTEST_AND_RUN(TpetraCrsGraph_isIdenticalTo)
  !TODO-FIX ADD_SUBTEST_AND_RUN(TpetraCrsGraph_replaceDomainMapAndImporter)
```

Tests that need improvement
```
    fresult = Graph1%isIdenticalTo(Graph2)
    fresult = Graph%getImporter(); TEST_IERR()
    fresult = Graph%getExporter(); TEST_IERR()
    fresult = Graph%description(); TEST_IERR()
    call Graph1%swap(Graph2); TEST_IERR()
```

Tests that fail if TEST_THROW is used
```
    !TEST_THROW(call graph%setParameterList(fresult))
    !TEST_THROW(call Graph%getGlobalRowCopy(globalrow, indices(:), numindices))
    !TEST_THROW(call Graph%getLocalRowCopy(localrow, indices(:), numindices))
    !TEST_THROW(call Graph%replaceColMap(newcolmap))
    !TEST_THROW(call Graph%removeEmptyProcessesInPlace(Map))
    !TEST_THROW(call Graph%computeGlobalConstants(farg))
```

To see the problems with `TEST_THROW` just uncomment out one of the above methods.

**Comment by aprokop** [comment]
First of all, there was a problem in some of the macros, producing wrong messages like `Test NAME FAILED!` without the proper name. I pushed the fix into the brunch. 

Second, I started looking at the `TEST_THROW` issues that you posted, and I think they are actually valid failures:
- `TEST_THROW(call Graph%getGlobalRowCopy(globalrow, indices(:), numindices))`
  This surprised me too. Apparently, it returns an empty row even if global index is invalid. It's undocumented in Doxygen, but is in the code.
- `TEST_THROW(call Graph%getLocalRowCopy(localrow, indices(:), numindices))`
  same as above

The rest `TEST_THROW` I'm not sure why you think they should throw.

**Comment by kruger** [comment]
I can't get your recent changes to compile.  I get errors with both flang on the docker image and gfortran.  They both give errors of the form
```
/scr_phi/kruger/triroot/trilinos/packages/ForTrilinos/src/teuchos/test/test_teuchos_comm.F90:24:88:

   ADD_SUBTEST_AND_RUN(TeuchosComm_Basic)
                                                                                        1
Error: Syntax error in argument list at (1)
```

Is there a flag or something else I am missing?  I'm unclear what this is.

Regarding your other points:  I'm a bit confused as to why the compiler isn't catching this for me when I don't have the `TEST_THROW`.  I'm going to investigate the use of bounds checking flags tomorrow.

I updated my swig today while doing some of this testing and am getting other errors.  I'll be talking to @sethrj on slack and show him my errors.

**Comment by sethrj** [comment]
@kruger I'm guessing that the error is because the macro expands to more than 130 (or whatever the default fortran limit) characters, and it truncates the rest of the expanded macro. You'll have to use a compiler option such as `-ffree-line-length-none` (for GCC) or we'll have to switch to a non-C preprocessor (like pfUnit, which preprocesses using python).

**Comment by kruger** [comment]
Thanks for the comments @sethrj but these aren't compile time errors, they are runtime errors.

**Comment by kruger** [comment]
@sethrj  You are correct.  That allows me to use the latest version, and then I have the other issues.  Thanks.

**Comment by aprokop** [comment]
Yes, I may have been too hasty in changing Tim's approach without thinking why it was there. Weirdly, it builds for me using Flang in the container... probably should not have been.

**Comment by tjfulle** [comment]
Hey all, sorry for the radio silence - I’ve been at DisneyLand with my kids.  I’ll be back on Monday morning.  @aprokop, what changes were made to the test macros that @kruger mentioned?

**Comment by aprokop** [comment]
@tjfulle I replaced some `"CODE"` and `"NAME"` to `#CODE` and `#NAME` without thinking about character limit.

**Comment by tjfulle** [comment]
I recall removing the `#` because of the line length limit.  I should have added a comment.  It may be possible to have a conditional macro that provides a different expansion for gnu and pgi (intel too?) since they allow longer line lengths (with the right compiler flags)

**Comment by sethrj** [comment]
I had a similar line length problem with the simple FASSERT macros I used in SWIG. I ended up just removing the code/line/name substitutions.

**Comment by kruger** [comment]
With my recent push to PR #262 the following tests are not working:

- TpetraCrsGraph_isIdenticalTo
- TpetraCrsGraph_replaceDomainMapAndImporter
- TpetraCrsMatrix_getAllValues
- TpetraCrsMatrix_setAllValues

I think it is just my confusion even after reading the documentation.  A Tpetra expert could probably quickly fix. 

Here are methods that probably need improvement (covered, but perhaps not robustly enough):

-TpetraCrsGraph:
    + getImporter
    + getExporter
    + getGlobalRowCopy
    + getLocalRowCopy

-TpetraCrsMatrix:
    + getGlobalRowCopy
    + getLocalRowCopy

**Comment by aprokop** [comment]
> With my recent push to PR #262 the following tests are not working:
>     * TpetraCrsGraph_isIdenticalTo

Can't reproduce, it's passing for me.

>     * TpetraCrsGraph_replaceDomainMapAndImporter
The following patch fixed it for me:
```diff
@@ -518,7 +518,6 @@ contains
     Graph1 = TpetraCrsGraph(Map, Map, izero, TpetraDynamicProfile)
     Graph2 = TpetraCrsGraph(Map, Map, izero, TpetraDynamicProfile)
 
-    !fresult=Graph1%isIdenticalTo(Graph2)
     TEST_ASSERT(Graph1%isIdenticalTo(Graph2))
 
     call Graph1%release()
@@ -1403,13 +1402,14 @@ contains
     num_global = 4*comm%getSize()
     Map = TpetraMap(num_global, comm); TEST_IERR()
     Graph = TpetraCrsGraph(Map, Map, izero, TpetraDynamicProfile)
+    call Graph%fillComplete()
 
     ! Make a new collmap
     num_global = 5*comm%getSize()
     newdomainmap = TpetraMap(num_global, comm); TEST_IERR()
 
-    !crash newimporter = TpetraImport(); TEST_IERR()
-    !call Graph%replaceDomainMapAndImporter(newdomainmap, newimporter); TEST_IERR()
+    newimporter = TpetraImport(newdomainmap, Graph%getColMap()); TEST_IERR()
+    call Graph%replaceDomainMapAndImporter(newdomainmap, newimporter); TEST_IERR()
 
     call newimporter%release(); TEST_IERR()
     call Graph%release()
```
According to [Doxygen documentation](https://trilinos.org/docs/dev/packages/tpetra/doc/html/classTpetra_1_1CrsGraph.html#ae117208949de5516ae8841f51a510339), the Graph must be fill complete. I don't understand at the moment why one is able to call `TpetraImport()` with no arguments. But providing the right constructor for it fixes the test.

The following patch makes it run (but does not test the result):
```diff
```

>     * TpetraCrsMatrix_getAllValues

This function returns copies of CRS arrays to the local (to process) matrix. The original Tpetra function returns pointers to internal structure, but as we want to make indices 1-based, we return the updated copies.

```diff
@@ -1021,35 +1021,31 @@ contains
   FORTRILINOS_UNIT_TEST(TpetraCrsMatrix_getAllValues)
     type(TpetraMap) :: rowmap, colmap
     type(TpetraCrsMatrix) :: Mat
-    integer :: numrow, numcol, nument, irow
-    integer(global_ordinal_type), dimension(:), pointer :: gblrow
-    integer, dimension(:), pointer :: colind
-    real(scalar_type), dimension(:), pointer :: values
-    !integer(global_ordinal_type), dimension(:), allocatable :: gblrow
-    !integer, dimension(:), allocatable :: colind
-    !real(scalar_type), dimension(:), allocatable :: values
+    integer :: numrow, numnnz
+    integer(size_type), dimension(:), allocatable :: row_ptrs
+    integer(int_type), dimension(:), allocatable :: colind
+    real(scalar_type), dimension(:), allocatable :: values
 
     OUT0("Starting TpetraCrsMatrix_getAllValues!")
 
-    call TPetra_CrsMatrix_CreateTestMatrix(comm,Mat)
-    call Mat%fillComplete()   
+    call TPetra_CrsMatrix_CreateTestMatrix(comm, Mat)
+    call Mat%fillComplete()
 
-    !TODO: Just not sure how this is supposed to work and I find Tpetra's 
-    ! documentation and test example confusing
-    numrow = Mat%getGlobalNumRows()
-    numcol = Mat%getGlobalNumCols()
-    nument = Mat%getGlobalNumEntries()
-    !WRITE(*,*) numrow, numcol, nument
-    !allocate(gblrow(numrow), colind(numcol), values(nument))
-    !rowmap = Mat%getRowMap()
-    !do irow=1,nument
-    !   gblrow(irow) = rowmap%getGlobalElement(irow)
-    !enddo
-    !TEST_THROW(call Mat%getAllValues(gblrow, colind, values))
-    call Mat%getAllValues(gblrow, colind, values)
+    numrow = Mat%getNodeNumRows()
+    numnnz = Mat%getNodeNumEntries()
+    allocate(row_ptrs(numrow+1))
+    allocate(colind(numnnz))
+    allocate(values(numnnz))
+
+    call Mat%getAllValues(row_ptrs, colind, values)
+
+    ! CHECK results
+    write(*,*) row_ptrs
+    write(*,*) colind
     write(*,*) values
 
-    !deallocate(gblrow, colind, values)
+    deallocate(row_ptrs, colind, values)
+
     call rowmap%release()
     call Mat%release()
```

>     * TpetraCrsMatrix_setAllValues

Similar to `getAllValues`, you need to create three arrays: one containing row pointers, also considered as offsets (this should be of size number of rows in the local matrix + 1), one containing (local) column indices (of size number of nonzeros you want to insert), and one containing values (of same size as the second array).

**Comment by aprokop** [comment]
Fixed by #262.
